### PR TITLE
Extract the HTTP request task from Hub

### PR DESCRIPTION
**Pull request by pragmader**
* This also involves extracting the token creator from Hub
* Also, moved the retention and SQL jobs to `handlers` package
* Fixed a go routine leak in the HTTP task handler's code

### CODE REVIEW

File: pkg/queue/handlers/http_request.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+	if err != nil {
+		return err
+	}
+
+	// the task would time out if the heartbeat was not sent for 30 seconds
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+	go func() {
+		for {
+			select {
```
Comment (pragmader) [inline pkg/queue/handlers/http_request.go:207]: Not having this select and ranging over the `ticket.C` channel caused the go routine leak detection. :man_shrugging:

File: pkg/queue/handlers/http_request.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+	RequestResponse HTTPRequestStage = "response"
+)
+
+// HTTPRequestProgress describes the progress of the HTTP request task stored during
+// the heartbeat handling
+type HTTPRequestProgress struct {
+	// Stage is the current stage of the HTTP request task
+	Stage HTTPRequestStage `json:"stage,omitempty"`
+	// Duration of the HTTP request in milliseconds
+	Duration *int64 `json:"duration,omitempty"`
```
Comment (trusch) [inline pkg/queue/handlers/http_request.go:62]: can this be a time.Duration?

Comment (trusch) [reply pkg/queue/handlers/http_request.go:62]: you get nice formatting and clearer semantics

File: pkg/queue/handlers/http_request.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+// HTTPRequestTaskSpec describes the specification of the HTTP request task
+type HTTPRequestTaskSpec struct {
+	// Method to use for the HTTP request
+	Method string `json:"method"`
+	// URL is the target URL for the request.
+	// Must be an absolute URL that contains the scheme and the host components.
+	URL string `json:"url"`
+	// RequestBody to send
+	RequestBody string `json:"requestBody"`
```
Comment (trusch) [inline pkg/queue/handlers/http_request.go:34]: Perhaps it makes sense to use []byte for the request body and the ResponseBody below. That way the handler is also usable for cases where binary data is submitted that is not valid utf8

File: pkg/queue/handlers/http_request.go (hunk 2 above)
Comment (pragmader) [reply pkg/queue/handlers/http_request.go:62]: good point, will do.

File: pkg/queue/handlers/http_request.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+	if spec.RequestBody != "" {
+		payload = strings.NewReader(spec.RequestBody)
+	}
+
+	req, err := http.NewRequest(spec.Method, spec.URL, payload)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("User-Agent", "Contiamo Hub HTTP Request Task")
```
Comment (trusch) [inline pkg/queue/handlers/http_request.go:139]: here is a mention of "hub"

File: pkg/queue/handlers/http_request.go (hunk 3 above)
Comment (pragmader) [reply pkg/queue/handlers/http_request.go:34]: this task handlers supports only `application/json` as a response, you can find the check in the code.

File: pkg/queue/handlers/http_request.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+	}()
+
+	resp, err := h.client.Do(req.WithContext(ctx))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	contentType := resp.Header.Get("content-type")
+	if !strings.Contains(contentType, "application/json") {
```
Comment (trusch) [inline pkg/queue/handlers/http_request.go:188]: why is this check needed?

File: pkg/queue/handlers/http_request.go (hunk 1 above)
Comment (trusch) [reply pkg/queue/handlers/http_request.go:207]: Strange. I mean you stop the ticker explicitly oO

File: pkg/queue/handlers/http_request.go (hunk 5 above)
Comment (pragmader) [reply pkg/queue/handlers/http_request.go:188]: Please read the implementation to the end. We support partial results which are new-line separated JSON objects, this task handlers supports ONLY JSON responses.

File: pkg/queue/handlers/http_request.go [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+		}
+	}()
+
+	decoder := json.NewDecoder(resp.Body)
+	for decoder.More() {
+		err = ctx.Err()
+		if err != nil {
+			return err
+		}
+		var m json.RawMessage
```
Comment (trusch) [inline pkg/queue/handlers/http_request.go:225]: that makes no sense to me. Decoding into a RawMessage means copying the data and nothing more. Perhaps it checks if its valid json, but nothing more. Why not just ignore the content type and pass through the body untouched?

**Review by trusch — CHANGES_REQUESTED** [review]

File: pkg/queue/handlers/http_request.go (hunk 6 above)
Comment (trusch) [reply pkg/queue/handlers/http_request.go:225]: got it, only json is supported.

**Review by trusch — APPROVED** [review]
Like discussed in slack, this looks really good except the name which is a little unspecific. Given that :+1:

**Review by trusch — APPROVED** [review]

**Pull request merged**